beacon/blsync/block_sync.go

```diff
@@ -117,15 +118,32 @@ func (s *beaconBlockSync) updateEventFeed() {
 	if !ok {
 		return
 	}
-	finality, ok := s.headTracker.ValidatedFinality() //TODO fetch directly if subscription does not deliver
-	if !ok || head.Header.Epoch() != finality.Attested.Header.Epoch() {
-		return
-	}
+
 	validatedHead := head.Header.Hash()
 	headBlock, ok := s.recentBlocks.Get(validatedHead)
 	if !ok {
 		return
 	}
+
+	var finalizedHash common.Hash
+	if finality, ok := s.headTracker.ValidatedFinality(); ok {
+		if he, fe := head.Header.Epoch(), finality.Attested.Header.Epoch(); he == fe {
+			finalizedHash = finality.Finalized.PayloadHeader.BlockHash()
+		} else {
+			if he < fe {
+				return // finality proof arrived earlier, wait for head update
+			}
+			if he == fe+1 {
+				if parent, ok := s.recentBlocks.Get(head.Header.ParentRoot); !ok ||
+					parent.Slot()/params.EpochLength == fe {
+					return // head is at first slot of next epoch, wait for finality update
+					//TODO trt to fetch finality update directly if subscription does not deliver
+				}
+
+			}
+		}
```

Comment:
```suggestion
		he := head.Header.Epoch()
		fe := finality.Attested.Header.Epoch()
		switch {
		case he == fe:
			finalizedHash = finality.Finalized.PayloadHeader.BlockHash()
		case he < fe:
			return
		case he == fe+1:
			parent, ok := s.recentBlocks.Get(head.Header.ParentRoot)	
			if !ok || parent.Slot()/params.EpochLength == fe {
				return // head is at first slot of next epoch, wait for finality update
				//TODO trt to fetch finality update directly if subscription does not deliver
			}
		}
```

Reply:
Nicer indeed, thanks. Committed and rebased on master.